### ISSUE TITLE
Not installable on Heroku since 1.5.10 release

### ISSUE THREAD

**Issue by wichert** [Toblerity/Shapely#322]
Shapely 1.5.9 installs correctly on Heroku, but later versions fail. The seems to be caused by a combination of how you have to install extra libraries on Heroku, and how  Shapely tries to find libgeos.

Extra software on Heroku can be installed using [heroku-buildpack-apt](https://github.com/heroku/heroku-buildpack-apt). This does something slightly odd: it downloads a deb package and unpacks it in ``/app/.apt``. It then sets a bunch of environment variables (INCLUDE_PATH, LIBRARY_PATH, LD_LIBRARY_PATH, PATH) to make sure the installed files are found normally. This does not work with current Shapely 1.5.11 and 1.5.12:

```
remote:        Collecting Shapely==1.5.12 (from -r requirements.txt (line 29))
remote:          Downloading Shapely-1.5.12.tar.gz (174kB)
remote:            Complete output from command python setup.py egg_info:
remote:            Traceback (most recent call last):
remote:              File "<string>", line 20, in <module>
remote:              File "/tmp/pip-build-k65yk40p/Shapely/setup.py", line 38, in <module>
remote:                from shapely.libgeos import geos_version_string, geos_version, \
remote:              File "/tmp/pip-build-k65yk40p/Shapely/shapely/libgeos.py", line 148, in <module>
remote:                fallbacks=['libgeos_c.so.1', 'libgeos_c.so'])
remote:              File "/tmp/pip-build-k65yk40p/Shapely/shapely/libgeos.py", line 143, in load_dll
remote:                libname, fallbacks))
remote:            OSError: Could not find library geos_c or load any of its variants ['libgeos_c.so.1', 'libgeos_c.so']
```

Shapely 1.5.10 failed differently:

```
remote:        Collecting Shapely==1.5.10 (from -r requirements.txt (line 29))
remote:          Downloading Shapely-1.5.10.tar.gz (173kB)
remote:            Complete output from command python setup.py egg_info:
remote:            Traceback (most recent call last):
remote:              File "<string>", line 20, in <module>
remote:              File "/tmp/pip-build-ehuoc1fy/Shapely/setup.py", line 34, in <module>
remote:                from packaging.version import Version
remote:            ImportError: No module named 'packaging'
remote:            
remote:            ----------------------------------------
remote: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-ehuoc1fy/Shapely
```

**Comment by sgillies** [comment]
@wichert we've changed a few things since 1.5.9 but we're still using `ctypes.util.find_library()` to find the .so. Heroku's environment shouldn't be a problem. I'll take a look in our setup.py, I think that's where the problem lies.

**Comment by sgillies** [comment]
@wichert would you be willing to try pip installing from the issue325 branch as a test? 

`pip install git+https://github.com/Toblerity/Shapely.git@issue325#egg=Shapely`

I'm working on the fix at https://github.com/Toblerity/Shapely/pull/327

**Comment by revmischa** [comment]
I'm using 1.5.13 and heroku-buildpack-apt, installing libgeos-c1
I get the following error:

```
remote:        Collecting Shapely==1.5.13 (from -r requirements.txt (line 20))
remote:          Downloading Shapely-1.5.13.tar.gz (176kB)
remote:            Complete output from command python setup.py egg_info:
remote:            cannot open shared object from 'geos-config --clibs': '-L/usr/lib -lgeos_c'
remote:            there were 2 guess(es) for this path:
remote:            	/usr/lib/libgeos_c.so
remote:            	/usr/lib/libgeos_c.so.1
remote:            Failed `CDLL(libgeos_c.so.1)`
remote:            Failed `CDLL(libgeos_c.so)`
remote:            Traceback (most recent call last):
remote:              File "<string>", line 20, in <module>
remote:              File "/tmp/pip-build-HndGSD/Shapely/setup.py", line 38, in <module>
remote:                from shapely._buildcfg import geos_version_string, geos_version, \
remote:              File "shapely/_buildcfg.py", line 167, in <module>
remote:                fallbacks=['libgeos_c.so.1', 'libgeos_c.so'])
remote:              File "shapely/_buildcfg.py", line 161, in load_dll
remote:                libname, fallbacks or []))
remote:            OSError: Could not find library geos_c or load any of its variants ['libgeos_c.so.1', 'libgeos_c.so']
```

**Comment by sgillies** [comment]
Souldn't your heroku buildpack installation have put geos libs at something like `$HOME/.apt/usr/lib` and `geos-config` program at `$HOME/.apt/usr/bin`? Looks like a geos-config program at /usr/bin is being found instead.

A `GEOS_CONFIG` environment variable can be used to tell Shapely to use a specific geos-config program when installing. Something like `GEOS_CONFIG=$HOME/.apt/usr/bin/geos-config pip install shapely==1.5.13` could help. Can you try that?

**Comment by revmischa** [comment]
I would be happy to. How do I instruct heroku to do that?
My project is at https://github.com/revmischa/dancar

**Comment by sgillies** [comment]
@revmischa I wish I could tell you. I'm not familiar with Heroku. I'm guessing that https://devcenter.heroku.com/articles/config-vars#setting-up-config-vars-for-a-deployed-application could help you.

**Comment by sgillies** [comment]
@revmischa any good news? I checked DANCAR (and laughed) and didn't see any updates since 10/30.

**Comment by revmischa** [comment]
Sadly no :(
Heroku support was unsatisfactory

**Comment by revmischa** [comment]
It appears that libgeos is installed to:
/app/.apt/usr/lib/libgeos_c.so
/app/.apt/usr/lib/libgeos-3.4.2.so
/app/.apt/usr/lib/libgeos_c.so.1.8.2
/app/.apt/usr/lib/libgeos_c.a
/app/.apt/usr/lib/libgeos_c.so.1

but geos-config returns the wrong path:
```
~ $ /app/.apt/usr/bin/geos-config  --clibs
-L/usr/lib -lgeos_c
```

**Comment by sgillies** [comment]
@revmischa Interesting! I can work with this. Would be nice for `geos-config` to yield the proper library path on Heroku, I wonder if the team is aware of this issue.

**Comment by wichert** [comment]
Heroku does that on purpose: they don't give you root access to their containers, but in order to allow you to install packages they use a wrapper which downloads a package, unpacks it and then copies the files to `/app/.apt` (and skipping any package configuration steps). The result is that are likely to get wrong paths from anything pkgconfig-related or everything else that assumes standard paths.

**Comment by alexjg** [comment]
Has anyone found a solution to this? I'm running into the same problem.

**Comment by revmischa** [comment]
I implemented it in python https://github.com/revmischa/dancar/blob/master/dancar/models.py#L10

**Comment by ezovski** [comment]
@alexjg I'm still testing it out, but it seems like adding the buildpack at https://github.com/cyberdelia/heroku-geo-buildpack/ builds geos and includes it in an expected way.

My Flask app seems to be successfully using it on Heroku. There are additional instructions for using it with Django at https://devcenter.heroku.com/articles/postgis.

**Comment by revmischa** [comment]
Okay. I'm using Flask and SQLAlchemy. All I needed Shapely for was converting the native lng/lats between pg and postgres. This code does it just fine, albiet with a round trip query https://github.com/revmischa/dancar/blob/master/dancar/models.py#L10
So now I don't need Shapely.

**Comment by sgillies** [comment]
Resolved by #419.

**Comment by MartinoMensio** [comment]
Just in case someone else is having still this problem, I found [there](https://gis.stackexchange.com/questions/62163/install-shapely-with-geos-on-heroku) that you can simply set this heroku environment variable: `heroku config:set BUILD_WITH_GEO_LIBRARIES=1` and build will succeed.
